File: Go/Day03.go
package main

import AH "./adventhelper"

func bitCount(m map[string]bool, size int) []int {
	counts := make([]int, size)

	for s, _ := range(m) {
		for i, r := range s {
			if r == '1' {
				counts[i] += 1
			}
  	}
	}

	return counts
}

func GammaEpsilon(m map[string]bool, size int) (gamma int, epsilon int) {
	total := len(m)
	mostCommon := bitCount(m, size)

	for _, b := range mostCommon {
		gamma <<= 1
		epsilon <<= 1

		if 2*b >= total {
			gamma += 1
		} else {
			epsilon += 1
		}
	}

	return
}

func stringToBin(s string) (bin int) {
	for _, r := range s {
		bin <<= 1

		if r == '1' {
			bin += 1
		}
	}	

	return
}

func setReduce(m map[string]bool, size int, good bool) int {
	for i := 0; i < size; i++ {
		total := len(m)
		mostCommon := bitCount(m, size)
		v := mostCommon[i]

		r := 'X'
		if 2*v >= total {
			r = '1'
		} else {
			r = '0'
		}

		for k, _ := range(m) {
			if ((rune(k[i]) != r) == good) { 
				delete(m, k)
			}
		}

		if len(m) == 1 {
			for k, _ := range(m) {
				return stringToBin(k)
			}
		}
	}

	return -1
}

func Oxygen(m map[string]bool, size int) (o2 int, co2 int) {
	o2Good  := make(map[string]bool)
	co2Good := make(map[string]bool)

	for k, v :=  range m {
		o2Good[k] = v
		co2Good[k] = v
	}

	// do these sperately
	o2  = setReduce(o2Good, size, true)
	co2 = setReduce(co2Good, size, false)

	return
}

func main() {
	bins, _ := AH.ReadStrFile("../input/input03.txt")
	binSet := make(map[string]bool)
	for _, v := range bins {
		binSet[v] = true
	}

	g, e := GammaEpsilon(binSet, len(bins[0]))
	o2, co2 := Oxygen(binSet, len(bins[0]))

	AH.PrintSoln(3, g * e, o2 * co2)

	return
}

File: Haskell/Day03.hs
import AdventHelper

bitCount :: [String] -> [Int]
bitCount ss = foldl1 (zipWith (+)) $ map singleCount ss
  where singleCount s = map (\c -> if' (c == '1') 1 0) s

toBin :: [Bool] -> Int
toBin b = foldl toBin' 0 b
  where toBin' x b = 2 * x + (if' b 1 0)

stringToBin :: String -> Int
stringToBin ss = foldl stringToBin' 0 ss
  where stringToBin' x c = 2 * x + (if' (c == '1') 1 0)

reduce :: Bool -> Int -> [String] -> String
reduce _ _ [s] = s
reduce b n ss = reduce b (n+1) ss'
  where bc = bitCount ss
        q = (2 * (bc!!n)) >= (length ss)
        t = if' q '1' '0'
        ss' = filter (\w -> (w!!n /= t) == b) ss

main = do
  putStrLn "Day 3"
  f <- readFile "../input/input03.txt"
  let ss = lines f

  let gamma = toBin $ map (\x -> 2*x >= length ss) $ bitCount ss
  let epsilon = toBin $ map (\x -> 2*x < length ss) $ bitCount ss
  
  let o2 = stringToBin $ reduce True 0 ss
  let co2 = stringToBin $ reduce False 0 ss

  printSoln 1 $ (gamma * epsilon)
  printSoln 2 $ (o2 * co2)
